packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/implementations/method_channel_document_reference.dart

```diff
@@ -0,0 +1,32 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart' show visibleForTesting;
+
+import '../interfaces/document_reference_platform.dart';
+
+class MethodChannelDocumentReference extends DocumentReferencePlatform {
+  MethodChannelDocumentReference() {
+    channel.setMethodCallHandler(_callHandler);
+  }
+
+  @visibleForTesting
+  static const MethodChannel channel = MethodChannel(
```

Comment:
in the original implementation I see that the channel is initialized with `FirestoreMessageCodec`. You probably need to move `FirestoreMessageCodec` to this package and make sure the channels are initialized with it.

Reply:
Thanks, I think I didn't realize this. I'll get this done!

Reply:
Oh my, this codec complicates things 😭

Reply:
Because of how the Codec operates (it basically uses the frontend of the plugin to decode some data types), I changed the initialization code of the MethodChannel Implementations to allow to pass a codec in; that way we can pass a codec with access to "userland" from the plugin, without having to migrate EVERY data type defined in the plugin (I tried, it was painful, and still didn't solve my issue).

This is [the problematic bit](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_firestore/cloud_firestore/lib/src/firestore_message_codec.dart#L91-L96).

Reply:
The code above probably explains why there's so much "static" code, and initialization guards in different places of the plugin; since we're creating new instances of the plugin all the time the message codec needs to hydrate a serialized DocumentReference!

Fixing this is *very* involved; we may be able to inject a factory function so the codec doesn't need to create instances in there, but maybe this particular part of the plugin could benefit from some extra design work (I didn't find any other place where layers are pierced like this, so far)

Reply:
Is hydrating a serialized DocumentReference expensive? It's not clear to how much overhead is involved in that, compared to the general overhead of moving serialized snapshots in and out of the Dart VM.

Reply:
Currently, it is *not* very expensive because:

* There's a bunch of init guards/static code that prevents multiple initialization
* We hydrate the moment we see a DcoumentReference that needs to be hydrated.

However, logic-wise, it **is** expensive, because the deserialization needs to be able to basically instantiate a whole new plugin. 

If we want to break that up, we'd probably need to hoist all the data types to the platform interface. At that moment, to maintain current feature parity, we'd need hydration on getters (or similar). The cost of that hydration would depend on the complexity of the returned data.

(Or maybe there's something I'm not seeing, there's a chance I've been focusing on this code for too long already :P)